File: segundoPeriodo/AEDS-2/estudos/hash/hashDiretoRehash/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct Hash {
    int *tabela, tam;
} Hash;

Hash *criarHash(int x) {
    Hash *nova = (Hash*)malloc(sizeof(Hash));
    nova->tabela = (int*)malloc(x * sizeof(int));
    nova->tam = x;
    for(int i = 0; i < nova->tam; i++) {
        nova->tabela[i] = -1;
    }
    return nova;
}

int h(Hash *hash, int x) {
    return x % hash->tam;
}
int res(Hash *hash, int x) {
    return ++x % hash->tam;
}

void inserir(Hash *hash, int x) {
    if(x != -1) {
        int num = h(hash, x);
        if(hash->tabela[num] == -1) hash->tabela[num] = x;
        else {
            int rehash = res(hash, x);
            if(hash->tabela[rehash] == -1) hash->tabela[rehash] = x;
            else printf("Nao eh possivel inserir o numero %d\n", x);
        }
    }
}

int remover(Hash *hash, int x) {
    int resp = -1;
    if(x != -1) {
        int num = h(hash, x);
        if(hash->tabela[num] == x) {
            hash->tabela[num] = -1;
            resp = x;
        }
        else {
            int rehash = res(hash, x);
            if(hash->tabela[rehash] == x) {
                hash->tabela[rehash] = -1;
                resp = x;
            }
        }
    }
    return resp;
}

bool pesquisarAux(Hash *hash, int x) {
    bool resp = false;
    if(x != -1) {
        int num = h(hash, x);
        if(hash->tabela[num] == x) {
            resp = true;
        }
        else {
            int rehash = res(hash, x);
            if(hash->tabela[rehash] == x) {
                resp = true;
            }
        }
    }
    return resp;
}
void pesquisar(Hash *hash, int x) {
    if(pesquisarAux(hash, x)) {
        printf("Elemento %d encontrado\n", x);
    }
    else {
        printf("Elemento %d nao encontrado\n", x);
    }
}

int mostrar(Hash *hash) {
    for(int i = 0; i < hash->tam; i++) {
        printf("%d ", hash->tabela[i]);
    }
    printf("\n");
}

void freeHash(Hash *hash) {
    free(hash->tabela);
    free(hash);
}

int main() {
    Hash *h = criarHash(7);
    inserir(h, 5);
    inserir(h, 7);
    inserir(h, 14);
    inserir(h, 21);
    inserir(h, 19);
    inserir(h, 1);
    mostrar(h);
    pesquisar(h, 19);
    remover(h, 19);
    mostrar(h);
    pesquisar(h, 19);
    freeHash(h);
    return 0;
}

File: segundoPeriodo/AEDS-2/estudos/hash/hashDiretoRehash/java/Main.java
class Hash {
    private int tabela[];
    private int tam;

    public Hash(int x) {
        tabela = new int[x];
        tam = x;
        for(int i = 0; i < tam; i++) {
            tabela[i] = -1;
        }
    }

    private int h(int x) {
        return x % tam;
    }
    private int reh(int x) {
        return ++x % tam;
    }

    public void inserir(int x) {
        if(x != -1) {
            int hash = h(x);
            if(tabela[hash] == -1) tabela[hash] = x;
            else {
                int rehash = reh(x);
                if(tabela[rehash] == -1) tabela[rehash] = x;
                else {
                    System.out.println("Nao eh possivel inserir o numero " + x);
                }
            }
        }
    }

    public int remover(int x) {
        int resp = -1;
        if(x != -1) {
            int hash = h(x);
            if(tabela[hash] == x) {
                tabela[hash] = -1;
                resp = x;
            }
        }
        else {
            int rehash = reh(x);
            if(tabela[rehash] == x) {
                tabela[rehash] = -1;
                resp = x;
            }
        }
        return resp;
    }

    public void pesquisar(int x) {
        if(pesquisarAux(x)) {
            System.out.println("Elemento " + x + " encontrado!");
        }
        else {
            System.out.println("Elemento " + x + " nao encontrado!");
        }
    }
    private boolean pesquisarAux(int x) {
        boolean resp = false;
        int hash = h(x);
        if(tabela[hash] == x) resp = true;
        else {
            int rehash = reh(x);
            if(tabela[rehash] == x) resp = true;
        }
        return resp;
    }

    public void mostrar() {
        for(int i = 0; i < tam; i++) {
            System.out.print(tabela[i] + " ");
        }
        System.out.println();
    }
}

public class Main {
    public static void main(String args[]) {
        Hash h = new Hash(7);
        h.inserir(5);
        h.inserir(7);
        h.inserir(14);
        h.inserir(21);
        h.inserir(19);
        h.inserir(1);
        h.mostrar();
        h.pesquisar(13);
    }
}